### PR TITLE
Move the functions of the Index Checker's MinLen Checker into the Value Checker

### PR DESCRIPTION
**Pull request by kelloggm**
This change makes `@MinLen(n)` an alias for `@ArrayLenRange(from = n)`, and updates the Index Checker to use this version instead of the origin MinLen Checker. The MinLen Checker is removed.

`@PolyMinLen` is removed; it will need to be replaced by `@PolyValue` when that is implemented (see typetools#1236).

I wanted to get this under review sooner rather than later, since it's a large change.

### CODE REVIEW

**Comment by smillst** [comment]
The test are failing because the jdk download has `@PolyMinLen` still.  I'll correct this.

File: framework/src/org/checkerframework/common/value/ValueTransfer.java [hunk 1]
```diff
@@ -925,7 +927,7 @@ private Range calculateRangeUnaryOp(
                 break;
             case GREATER_THAN_EQ:
                 thenRightRange = rightRange.refineLessThanEq(leftRange);
-                thenLeftRange = leftRange.refineGreaterThan(rightRange);
+                thenLeftRange = leftRange.refineGreaterThanEq(rightRange);
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/ValueTransfer.java]: Is this just a bug?

File: framework/src/org/checkerframework/common/value/ValueTransfer.java [hunk 2]
```diff
@@ -1210,4 +1212,28 @@ private void addAnnotationToStore(CFStore store, AnnotationMirror anno, Node nod
...
+     * Node, Node, CFAbstractValue, CFAbstractValue, boolean)}, which is called by {@link
+     * org.checkerframework.framework.flow.CFAbstractTransfer#visitCase(CaseNode, TransferInput)}.
+     * Therefore, we need to explicitly call the handling for binary comparison if we want to do the
+     * extra refinements that the Value Checker applies on switch statements.
+     */
+    @Override
+    public TransferResult<CFValue, CFStore> visitCase(
+            CaseNode n, TransferInput<CFValue, CFStore> p) {
+        TransferResult<CFValue, CFStore> transferResult = super.visitCase(n, p);
+        if (n.getSwitchOperand() instanceof FieldAccessNode) {
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/ValueTransfer.java]: Why just field accesses?

File: framework/src/org/checkerframework/common/value/ValueTransfer.java [hunk 3]
```diff
@@ -1210,4 +1212,28 @@ private void addAnnotationToStore(CFStore store, AnnotationMirror anno, Node nod
                 resultValues,
                 transferResult.getResultValue().getUnderlyingType());
     }
+
+    /**
+     * The value checker doesn't override {@link
+     * org.checkerframework.framework.flow.CFAbstractTransfer#strengthenAnnotationOfEqualTo(TransferResult,
+     * Node, Node, CFAbstractValue, CFAbstractValue, boolean)}, which is called by {@link
+     * org.checkerframework.framework.flow.CFAbstractTransfer#visitCase(CaseNode, TransferInput)}.
+     * Therefore, we need to explicitly call the handling for binary comparison if we want to do the
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/ValueTransfer.java]: Please rewrite to not use "we".

File: framework/src/org/checkerframework/common/value/qual/MinLen.java [hunk 4]
```diff
@@ -1,18 +1,14 @@
...
 
 import java.lang.annotation.ElementType;
 import java.lang.annotation.Target;
-import org.checkerframework.framework.qual.DefaultQualifierInHierarchy;
-import org.checkerframework.framework.qual.SubtypeOf;
 
 /**
  * The value of the annotated expression is a sequence containing at least the given number of
- * elements.
+ * elements. An alias for an {@link ArrayLenRange} annotation with only a {@code from} field.
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/qual/MinLen.java]: Please explain what value the "to" field of ArrayLenRange has.

File: framework/src/org/checkerframework/common/value/qual/MinLen.java [hunk 5]
```diff
@@ -1,18 +1,14 @@
...
  *
  * @checker_framework.manual #index-checker Index Checker
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/qual/MinLen.java]: This should be documented in the Value Checker -- not the Index Checker.

File: framework/src/org/checkerframework/common/value/qual/MinLenFieldInvariant.java [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package org.checkerframework.checker.index.qual;
+package org.checkerframework.common.value.qual;
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/qual/MinLenFieldInvariant.java]: The Javadoc (below) is out-of-date.  Also, would it make more sense for this to be changed to ArrayLenRangeFieldInvariant?  (This is something I could do.)

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java [hunk 7]
```diff
@@ -707,10 +707,13 @@ public AnnotationMirror getAnnotationFromJavaExpressionString(
 
         AnnotationMirror annotationMirror = null;
         if (CFAbstractStore.canInsertReceiver(expressionObj)) {
-            Value value = getStoreBefore(tree).getValue(expressionObj);
-            if (value != null) {
-                annotationMirror =
-                        AnnotationUtils.getAnnotationByClass(value.getAnnotations(), clazz);
+            Store store = getStoreBefore(tree);
```
Comment (smillst) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: This should be in a separate PR with a test case.  It's a general bug fix to a framework class.

File: framework/src/org/checkerframework/common/value/qual/MinLen.java [hunk 8]
```diff
@@ -1,18 +1,14 @@
...
  */
-@DefaultQualifierInHierarchy
-@SubtypeOf({})
 @Target({ElementType.TYPE_USE, ElementType.TYPE_PARAMETER})
 public @interface MinLen {
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/qual/MinLen.java]: Should we rename this?  It seems out of place with  `@ArrayLen` and `@ArrayLenRange`.  Maybe `@ArrayMinLen`?

File: docs/manual/index-checker.tex [hunk 9]
```diff
@@ -303,23 +303,27 @@ \section{Array minimum lengths\label{index-minlen}}
 annotated as
 \refqualclasswithparams{checker/index/qual}{LTLengthOf}{"arr"}.
 
+The MinLen type system is implemented in practice by the Constant Value Checker
```
Comment (smillst) [inline docs/manual/index-checker.tex]: `@MinLen` should be documented in the Value Checker section, not here.

File: framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java [hunk 10]
```diff
@@ -1184,6 +1245,20 @@ private AnnotationMirror convertIntValToDoubleVal(AnnotationMirror intValAnno) {
     }
 
     /**
+     * Builds an ArrayLenRange annotation with from equal to its argument. This is called by the
+     * clients that used the old MinLen Checker, and provides the same interface.
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java]: Don't reference the old MinLen Checker -- just document what this does.

File: framework/src/org/checkerframework/common/value/ValueTransfer.java (hunk 1 above)
Comment (kelloggm) [inline framework/src/org/checkerframework/common/value/ValueTransfer.java]: Yes, this is just a bug. I probably should have PR'd it separately; I can do that if you'd like.

File: docs/manual/index-checker.tex (hunk 9 above)
Comment (kelloggm) [inline docs/manual/index-checker.tex]: The Index Checker's manual section should at least include an explanation of how to use `@MinLen` and why a programmer needs it while using the Index Checker. I'm fine with moving the documentation of the annotation itself to the Value Checker's manual section, though.

File: checker/jdk/index/src/java/lang/String.java [hunk 11]
```diff
@@ -162,7 +162,7 @@ public String() {
      * @param  original
      *         A {@code String}
      */
-    public @PolyMinLen @PolySameLen String(@PolyMinLen @PolySameLen String original) {
+    public /*@ PolyMinLen*/ @PolySameLen String(/*@ PolyMinLen*/ @PolySameLen String original) {
```
Comment (smillst) [inline checker/jdk/index/src/java/lang/String.java]: `@ PolyMinLen` is the same as `@PolyMinLen`, so it's getting compiled into the code and causing the tests to fail.

File: framework/src/org/checkerframework/common/value/ValueTransfer.java (hunk 2 above)
Comment (kelloggm) [inline framework/src/org/checkerframework/common/value/ValueTransfer.java]: This problem only came up when trying to refine `array.length` field accesses, but it is more general, so the `if` guard can (and should) probably be removed.

File: framework/src/org/checkerframework/common/value/ValueTransfer.java (hunk 1 above)
Comment (smillst) [inline framework/src/org/checkerframework/common/value/ValueTransfer.java]: Please do.  Since this PR is likely to take a few days before it is merged, it would be good to have this fix change in master sooner.

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java (hunk 7 above)
Comment (kelloggm) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: Actually this should just be reverted; I didn't mean to include it in the PR (and thought I'd removed it). Sorry about that.

File: checker/jdk/index/src/java/lang/String.java (hunk 11 above)
Comment (kelloggm) [inline checker/jdk/index/src/java/lang/String.java]: Really? We use `/*@ AnnotationName*/` instead of `@AnnotationName` to comment out annotations throughout the Index Annotated JDK. If they're treated the same, I need to go through the whole Index JDK and change many things (especially the annotations on Strings, which I wrote but have since removed because we deferred String support).

File: framework/src/org/checkerframework/common/value/qual/MinLen.java (hunk 8 above)
Comment (kelloggm) [inline framework/src/org/checkerframework/common/value/qual/MinLen.java]: I prefer `@MinLen` still, but if you and @mernst prefer `@ArrayMinLen` or something else, we can change it. I think `@MinLen` is preferable because we'd like to eventually expand it to also be useful for Strings and possibly lists in the future.

File: checker/jdk/index/src/java/lang/String.java (hunk 11 above)
Comment (smillst) [inline checker/jdk/index/src/java/lang/String.java]: Try this:
```
void test() {
        @MinLen(4) String s = String.valueOf(true);
    }
```
A annotation must have `@Target(TYPE_USE)` for it to be read from comments.   `@IndexFor` doesn't have that so it is ignored if it is in comments (with or without the spaces).  That's why you didn't run into this problem before.

But yes, you should probably fix that.  And IndexFor (and related annotations) should be have `@Target({TYPE_USE, TYPE_PARAMETER})`

File: framework/src/org/checkerframework/common/value/qual/MinLen.java (hunk 8 above)
Comment (smillst) [inline framework/src/org/checkerframework/common/value/qual/MinLen.java]: >I think @MinLen is preferable because we'd like to eventually expand it to also be useful for Strings and possibly lists in the future.

At that time `@ArrayLen` and `@ArrayLenRange` would have to be changed, too.

Comment (mernst) [inline framework/src/org/checkerframework/common/value/qual/MinLen.java]: Good point.  I'm in favor of renaming them all now.  The longer we wait, the more annotated code will exist that has to be updated.

File: checker/jdk/index/src/java/lang/String.java (hunk 11 above)
Comment (kelloggm) [inline checker/jdk/index/src/java/lang/String.java]: I've issued a PR for this: typetools#1247

Comment (kelloggm) [inline checker/jdk/index/src/java/lang/String.java]: The changeset for this PR will be big until typetools#1247 is merged, because it includes all the changes to the JDK in typetools#1247. I did it programmatically over the whole Index JDK directory.

File: framework/src/org/checkerframework/common/value/qual/MinLen.java (hunk 8 above)
Comment (kelloggm) [inline framework/src/org/checkerframework/common/value/qual/MinLen.java]: Should that change be part of this PR? I think it would be more appropriate as a separate PR. I can do it, if you'd like.

Comment (smillst) [inline framework/src/org/checkerframework/common/value/qual/MinLen.java]: I agree, this should be a separate PR. (Probably, after this one is merged!)

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java (hunk 7 above)
Comment (smillst) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: I had to put this back in because without it there was a crash in when building the jdk.   It seems like correct code to me -- it just needs to be in a different PR.

File: framework/src/org/checkerframework/common/value/ValueTransfer.java [hunk 12]
```diff
@@ -1210,4 +1212,30 @@ private void addAnnotationToStore(CFStore store, AnnotationMirror anno, Node nod
...
+     * Node, Node, CFAbstractValue, CFAbstractValue, boolean)}, which is called by {@link
+     * org.checkerframework.framework.flow.CFAbstractTransfer#visitCase(CaseNode, TransferInput)}.
+     * Therefore, this needs to explicitly call the handling for binary comparison to do the extra
+     * refinements that the Value Checker applies on switch statements.
+     */
+    @Override
+    public TransferResult<CFValue, CFStore> visitCase(
+            CaseNode n, TransferInput<CFValue, CFStore> p) {
+        TransferResult<CFValue, CFStore> transferResult = super.visitCase(n, p);
+        // TODO: this should work for chars too, but it is causing a crash when building the
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/ValueTransfer.java]: I think the correct thing to do is make the Value Checker override strengthenAnnotationOfEqualTo.  If the types are numerical, call calculateBinaryComparison; otherwise, it should refine the left and right hand sides to the intersection of the known values.  I'll work on this.

File: framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java [hunk 13]
```diff
@@ -1506,21 +1581,87 @@ public static Range getRange(AnnotationMirror rangeAnno) {
...
+            return null;
+        }
+    }
+
+    /**
+     * Returns the minimum value of an integral type. This is used by the Index Checker's array
+     * bounds checking routines when determining the minimum length of an array from the annotation
+     * on the length of the array.
+     */
+    public Integer getMinLenValueFromLengthType(AnnotatedTypeMirror atm) {
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java]: Why `LengthType`?  Currently this is only used for array lengths, but it is (or should be) a general purpose routine that other checkers could use.  I would name it something like `getMinIntegralValue` and remove the mention of the Index Checker from the Javadoc.

Oh maybe it's `LengthType` because this method returns an Integer instead of a Long?  In either case, this should return a long and then the caller of the method can downcast as needed.

File: framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java [hunk 14]
```diff
@@ -1189,6 +1250,20 @@ private AnnotationMirror convertIntValToDoubleVal(AnnotationMirror intValAnno) {
     }
 
     /**
+     * Builds an ArrayLenRange annotation with {@code from} equal to its argument. The {@code to}
+     * field is set to Integer.MAX_VALUE, the default.
+     *
+     * @param val The value to use as the from in the ArrayLenRange annotation created. This value
+     *     must be non-negative.
+     */
+    public AnnotationMirror createMinLen(int val) {
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java]: I find this method name confusing since it is actually creating an `@ArrayLenRange`.  This method is only called once, so could you just replace that call with one to `createArrayLenRangeAnnotation(val, MAX_VALUE)`

File: framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java [hunk 15]
```diff
@@ -1506,21 +1581,87 @@ public static Range getRange(AnnotationMirror rangeAnno) {
...
+     * bounds checking routines when determining the minimum length of an array from the annotation
+     * on the length of the array.
+     */
+    public Integer getMinLenValueFromLengthType(AnnotatedTypeMirror atm) {
+        AnnotationMirror anm = atm.getAnnotationInHierarchy(UNKNOWNVAL);
+        if (AnnotationUtils.areSameByClass(anm, IntVal.class)) {
+            List<Long> possibleValues = getIntValues(anm);
+            Long minlen = Collections.min(possibleValues);
+            if (minlen < 0) {
+                return 0;
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java]: The caller should handle this case.

File: framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java [hunk 16]
```diff
@@ -1506,21 +1581,87 @@ public static Range getRange(AnnotationMirror rangeAnno) {
...
+            } else {
+                return minlen.intValue();
+            }
+        } else if (AnnotationUtils.areSameByClass(anm, IntRange.class)) {
+            Range range = getRange(anm);
+            if (range.from < 0) {
```
Comment (smillst) [inline framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java]: The caller should handle this case.

File: framework/src/org/checkerframework/common/value/qual/MinLenFieldInvariant.java (hunk 6 above)
Comment (smillst) [inline framework/src/org/checkerframework/common/value/qual/MinLenFieldInvariant.java]: Ping.

Comment (kelloggm) [inline framework/src/org/checkerframework/common/value/qual/MinLenFieldInvariant.java]: I'm not sure how to change this, since I don't really understand what it's doing in the first place. Should I change the name, but leave it so that it only accepts a lower bound (i.e. a MinLen)? That doesn't seem like the right thing to do, but I'm not sure what I should add to make it able to handle all of ArrayLenRange.

Comment (smillst) [inline framework/src/org/checkerframework/common/value/qual/MinLenFieldInvariant.java]: I updated the Javadoc.  For now, let's just leave it as is.  After this PR is merged, we can decide if we want to change it.

File: docs/manual/index-checker.tex [hunk 17]
```diff
@@ -303,23 +303,27 @@ \section{Array minimum lengths\label{index-minlen}}
 annotated as
 \refqualclasswithparams{checker/index/qual}{LTLengthOf}{"arr"}.
 
+The MinLen type system is implemented in practice by the Constant Value Checker,
```
Comment (smillst) [inline docs/manual/index-checker.tex]: MinLen is no longer a type system. Say something like, `@MinLen` is checked by the CVC.  Then maybe add something about how the other two array annotations can be used, too.  Maybe add an example using `@ArrayLen`.

File: docs/manual/index-checker.tex [hunk 18]
```diff
@@ -303,23 +303,27 @@ \section{Array minimum lengths\label{index-minlen}}
...
+using \<@ArrayLenRange> annotations (see \chapterpageref{constant-value-checker}).
+
 Array minimum lengths use the following type qualifiers
 (Figure~\ref{fig-index-array-types}):
 
 \begin{description}
```
Comment (smillst) [inline docs/manual/index-checker.tex]: `@MinLen` is documented in the Value Checker now, so remove this.

File: docs/manual/figures/minlen.svg [hunk 19]
```diff
@@ -176,7 +176,7 @@
       <tspan
          x="460.5"
          y="466.772"
-         id="tspan4346">@MinLenBottom</tspan>
+         id="tspan4346">@BottomVal</tspan>
```
Comment (smillst) [inline docs/manual/figures/minlen.svg]: MinLen is not a type system, so remove this figure.

File: framework/src/org/checkerframework/common/value/ValueTransfer.java (hunk 12 above)
Comment (smillst) [inline framework/src/org/checkerframework/common/value/ValueTransfer.java]: I've done this in https://github.com/typetools/checker-framework/pull/1251.  We can't merge this pull request until that one is merged.

File: framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java (hunk 7 above)
Comment (smillst) [inline framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java]: I create a PR, #1254, for this.

**Comment by kelloggm** [comment]
@smillst I think I've fixed all the documentation issues you pointed out yesterday. Can you review again?

**Comment by kelloggm** [comment]
Those tweaks look fine - thanks for catching those mistakes. Merging after the release on May 1 makes sense to me.

**Comment by smillst** [comment]
@kelloggm Could you please update the Javadoc in IndexChecker to reflect these changes?

**Comment by kelloggm** [comment]
@smillst Good catch. I've updated those docs.

File: checker/src/org/checkerframework/checker/index/IndexChecker.java [hunk 20]
```diff
@@ -3,43 +3,56 @@
...
  * described, and then what each checker requests from the checkers that run before it is described.
  * The Index Checker itself is just an alias for the Upper Bound Checker, which runs last.
  *
  * <p>The checkers run in this order:
  *
- * <p>SameLen Checker, MinLen Checker, Lower Bound Checker, Upper Bound Checker
+ * <p>Constant Value Checker, SameLen Checker, SearchIndex Checker, Lower Bound Checker, Upper Bound
+ * Checker
+ *
+ * <p>The Constant Value Checker has no dependencies, but it does trust Positive annotations from
```
Comment (smillst) [inline checker/src/org/checkerframework/checker/index/IndexChecker.java]: Technically, this paragraph should be added in https://github.com/kelloggm/checker-framework/pull/155, but it's fine with me if you want to leave it here, too.

**Comment by smillst** [comment]
IndexChecker Javadoc looks good.  Thanks!

File: checker/src/org/checkerframework/checker/index/IndexChecker.java (hunk 20 above)
Comment (kelloggm) [inline checker/src/org/checkerframework/checker/index/IndexChecker.java]: Oops, I forgot which PR this was! I have to go to class now, but I'll change it later today.

**Comment by kelloggm** [comment]
@smillst Now that the release has passed are we okay to merge this?
